### PR TITLE
WebSocket encryption

### LINKED ISSUE #2295: Encrypt WebSocket communication
Exchanged passwords should be encrypted.
~An outdated, basic implementation exists here: https://github.com/golemfactory/golem/commit/aafe14e78794e5f1d17689503d8936eef50a2143~

### PR DESCRIPTION
**Pull request by mfranciszkiewicz**
- generates and self-signs a certificate for TLS connections
- the self-signed certificate is trusted only by core
- forward secrecy is disabled by default due to long DHKE param generation
- crossbar certificate verification is optional in CLI (**RFC**)
- fixes startup errors in `golem.node.Node`

Related PR in `golem-electron`: https://github.com/golemfactory/golem-electron/pull/255 
Resolves #2295

### CODE REVIEW

File: golemcli.py [hunk 1]
```diff
@@ -7,6 +7,8 @@
 from multiprocessing import freeze_support
 
 # Export pbr version for peewee_migrate user
+from golem.rpc.common import CROSSBAR_HOST, CROSSBAR_PORT
```
Comment (maaktweluit) [inline golemcli.py]: would move this up 1 line to keep the comment and the os.env... together

Comment (mfranciszkiewicz) [reply golemcli.py]: thanks, my editor's import feature broke the comment

File: golem/node.py [hunk 2]
```diff
@@ -98,11 +99,17 @@ def _start_rpc(self) -> Deferred:
         self._reactor.addSystemEventTrigger("before", "shutdown", rpc.stop)
 
         # pylint: disable=protected-access
-        deferred = rpc._start_node(rpc.options, self._reactor)
+        deferred = rpc.start(self._reactor)
```
Comment (tworec) [inline golem/node.py:144]: 👍

File: golem/rpc/cert.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+        cert.set_issuer(cert_subject)
+        cert.set_pubkey(key)
+        cert.sign(key, 'sha1')
+
+        with open(output_path, "wb") as cert_file:
+            buffer = crypto.dump_certificate(crypto.FILETYPE_PEM, cert)
+            cert_file.write(buffer)
+        del cert
+
+    @classmethod
```
Comment (tworec) [inline golem/rpc/cert.py]: why not static? `cls` is not used

File: golem/rpc/cert.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        logger.info("Creating a self-signed certificate: %r", output_path)
+
+        cert = crypto.X509()
+        cert_subject = cert.get_subject()
+
+        cls._apply_to_subject(cert_subject, **entity)
+
+        cert.set_serial_number(random.randint(0, 10 * 10 ** 18))
+        cert.gmtime_adj_notBefore(0)
+        cert.gmtime_adj_notAfter(10 * 365 * 24 * 60 * 60)
```
Comment (tworec) [inline golem/rpc/cert.py:106]: 10y? it need's a comment

File: golem/rpc/session.py [hunk 5]
```diff
@@ -92,9 +103,29 @@ def cleanup(proto):
...
+        if self.address.ssl:
+            if self._cert_manager:
+                cert_data = self._cert_manager.read_certificate()
+                authority = twisted_ssl.Certificate.loadPEM(cert_data)
+            else:
+                authority = None
+
+            context_factory = optionsForClientTLS(X509_COMMON_NAME,
+                                                  trustRoot=authority)
+            self._client = SSL4ClientEndpoint(reactor,
```
Comment (tworec) [inline golem/rpc/session.py:115]: would be nice to have tests for this SSL client, especially with auth failure

**Review by tworec — APPROVED** [review]
lgtm

File: golem/rpc/session.py (hunk 5 above)
Comment (mfranciszkiewicz) [reply golem/rpc/session.py:115]: It logs and propagates errors to the returned Deferred object.

File: golem/rpc/cert.py (hunk 4 above)
Comment (mfranciszkiewicz) [reply golem/rpc/cert.py:106]: Done

**Pull request merged**